RFCs/rfc-0008-wallet-provider-api.md

```diff
@@ -0,0 +1,387 @@
+
+The remaining four events can be separated into two categories:
+
+- Changes to the Provider's ability to make RPC requests
+    - `connect`
+    - `disconnect`
+- Common Client and/or Wallet state changes that any non-trivial application must handle
+    - `chainChanged`
+    - `accountsChanged`
+
+### Security Considerations
+
+The Provider is intended to pass messages between a Mina Client and a Mina application. It is not responsible for private key or account management; it merely processes RPC messages and emits events. Consequently, account security and user privacy need to be implemented in middlewares between the Provider and its Mina Client. In practice, we call these middleware applications "Wallets," and they usually manage the user's private keys and accounts. The Provider can be thought of as an extension of the Wallet, exposed in an untrusted environment, under the control of some third party (e.g. a website).
+
+#### Handling Adversarial Behavior
+
+Since it is a JavaScript object, consumers can generally perform arbitrary operations on the Provider, and all its properties can be read or overwritten. Therefore, it is best to treat the Provider object as though it is controlled by an adversary. It is paramount that the Provider implementer protects the user, Wallet, and Client by ensuring that:
+
+- The Provider does not contain any private user data.
+- The Provider and Wallet programs are isolated from each other.
+- The Wallet and/or Client rate-limit requests from the Provider.
+- The Wallet and/or Client validate all data sent from the Provider.
+
+### Provider RPC Methods
+This section lists the standard RPC methods that can be supported by the Provider, along with their return types. Each method follows the `mina_{methodName}` naming convention.
+
+#### mina_accounts
+Returns a list of addresses owned by client.
+
+##### Parameters
+None
+
+##### Returns
+An array of hexadecimals as strings representing the addresses owned by the client.
+
+```typescript
+Provider.request({ 
+  method: 'mina_accounts'
+}): Promise<string[]>;
+```
+
+#### mina_chainId
+Returns the chain ID of the current network.
+
+##### Parameters
+None
+
+##### Returns
+A hexadecimal of the current chain ID.
+
+```typescript
+Provider.request({
+    method: 'mina_chainId'
+}): Promise<string>;
+```
+
+#### mina_sign
+Returns a signed message.
+
+##### Parameters
+`message`: [Required] a string
+
+##### Returns
+A signed message.
+
+```typescript
+type SignedMessage = {
+    publicKey: string;
+    data: string;
+    signature: {
+        field: string;
+        scalar: string;
+    };
+};
+
+Provider.request({
+    method: 'mina_sign',
+    params: { message: "hi bob" }
+}): Promise<SignedMessage>;
+```
+
+#### mina_signFields
+Returns a signed message.
+
+##### Parameters
+`fields`: [Required] an array of numbers or strings
+
+##### Returns
```

Comment:
The current return type of `[mina-signer].signFields` is the following, which is different from `[mina-signer].signMessage`
```typescript
type SignedFieldsData = {
  data: (string | number)[];
  publicKey:string;
  signature: string;
}
```

Reply:
And should we provide a method to `verifyFields` and `verifyMessage`?

Reply:
> And should we provide a method to verify signed fields and message ?

I'm happy to add in verification methods! I wonder what scenario they would be useful to a zkApp however, verification of a message or signed field might be an internal wallet feature rather than for a zkApp to request the verification of a wallet. A malicious wallet could of course choose not to provide the correct verification result!

Reply:
In fact, I just assumed that this scenario existed, but I haven't come across a place where I need to use this. It’s just that we tend to expose the `mina-signer` method to give zkApp greater freedom.

Reply:
I've updated the `SignedFieldsData` type.